odoo/tools/image.py

```diff
@@ -284,39 +307,38 @@ def image_get_resized_images(base64_source, return_big=False, return_medium=True
-    size_small = sizes.get(small_name, (64, 64))
+    size_big = sizes.get(big_name, IMAGE_BIG_SIZE)
+    size_medium = sizes.get(medium_name, IMAGE_MEDIUM_SIZE)
+    size_small = sizes.get(small_name, IMAGE_SMALL_SIZE)
     if isinstance(base64_source, str):
         base64_source = base64_source.encode('ascii')
     if return_big:
-        return_dict[big_name] = image_resize_image_big(base64_source, avoid_if_small=avoid_resize_big, size=size_big)
+        return_dict[big_name] = image_resize_image(base64_source, avoid_if_small=avoid_resize_big, size=size_big, preserve_aspect_ratio=preserve_aspect_ratio, upper_limit=upper_limit)
     if return_medium:
-        return_dict[medium_name] = image_resize_image_medium(base64_source, avoid_if_small=avoid_resize_medium, size=size_medium)
+        return_dict[medium_name] = image_resize_image(base64_source, avoid_if_small=avoid_resize_medium, size=size_medium, preserve_aspect_ratio=preserve_aspect_ratio, upper_limit=upper_limit)
     if return_small:
-        return_dict[small_name] = image_resize_image_small(base64_source, avoid_if_small=avoid_resize_small, size=size_small)
+        return_dict[small_name] = image_resize_image(base64_source, avoid_if_small=avoid_resize_small, size=size_small, preserve_aspect_ratio=preserve_aspect_ratio, upper_limit=upper_limit)
     return return_dict
 
-def image_resize_images(vals, big_name='image', medium_name='image_medium', small_name='image_small', sizes={}):
+
+def image_resize_images(vals, big_name='image', medium_name='image_medium', small_name='image_small', sizes={}, preserve_aspect_ratio=False, upper_limit=False):
     """ Update ``vals`` with image fields resized as expected. """
-    if vals.get(big_name):
-        vals.update(image_get_resized_images(vals[big_name],
-                        return_big=True, return_medium=True, return_small=True,
-                        big_name=big_name, medium_name=medium_name, small_name=small_name,
-                        avoid_resize_big=True, avoid_resize_medium=False, avoid_resize_small=False, sizes=sizes))
-    elif vals.get(medium_name):
-        vals.update(image_get_resized_images(vals[medium_name],
-                        return_big=True, return_medium=True, return_small=True,
-                        big_name=big_name, medium_name=medium_name, small_name=small_name,
-                        avoid_resize_big=True, avoid_resize_medium=True, avoid_resize_small=False, sizes=sizes))
-    elif vals.get(small_name):
-        vals.update(image_get_resized_images(vals[small_name],
-                        return_big=True, return_medium=True, return_small=True,
-                        big_name=big_name, medium_name=medium_name, small_name=small_name,
-                        avoid_resize_big=True, avoid_resize_medium=True, avoid_resize_small=True, sizes=sizes))
+    big_image = vals.get(big_name)
+    medium_image = vals.get(medium_name)
+    small_image = vals.get(small_name)
+
```

Comment:
no code for large image here ?

Reply:
As discussed, can't write image large on models that call this method but don't have the field.

Reply:
Finally I also added code for large_image there, with default return_large=False.

This is more consistent this way.